### ISSUE TITLE
Strange errors when used with ambassador(envoy) on kubernetes

### ISSUE THREAD

**Issue by vangheem** [Kludex/uvicorn#405]
I'm getting a lot of errors from starlette when putting it behind ambassador(envoy) on kubernetes.

503: upstream connect error or disconnect/reset before headers. reset reason: connection failure

Any idea what might be going wrong here?

**Comment by vangheem** [comment]
And when I get these errors, there are no logs in the uvicorn process.

**Comment by vangheem** [comment]
If anyone else is experiencing this issue, it seems to be related to http2 handling.

If you change the `http` setting in uvicorn to `h11`, you no longer see the issue.

I don't know if the issue stems from httptools or ambassador but in any case, http2 seems broken...

**Comment by vangheem** [comment]
spoke too soon... still seeing the issue.

**Comment by vangheem** [comment]
Final update: went with nginx side car to finally stop the nonsense.

I'll just close as it doesn't seem like there is much maintenance on this project.

FWIW, it also happened with hypercorn. So it might be some of the underlying asgi tooling at fault...

**Comment by lovelydinosaur** [comment]
This doesn't seem like something I'd be able to dig into.
We'd need to have something minimally reproducible in order to be able to help.
If the error reproduces on both `httptools` and `h11` implementations (and `hypercorn`) then it's likely not in related to the HTTP handling in uvicorn. (Perhaps you're getting connection errors during a startup phase or something like that?)

**Comment by anthonybishopric** [comment]
@vangheem @tomchristie I found this issue while debugging the same thing - in my case, the response body was empty but uvicorn/FastAPI was emitting a non-zero content length header, causing Envoy to think that there was more content at the time when the connection closed.

**Comment by lorenzobenvenuti** [comment]
I found this thread after googling the same error message. After some debugging, I was able to figure it out: in my case the issue was the TLS version. By default, Envoy has TLS max version set to 1.2 for upstream connections and this doesn't work with Uvicorn (although I can connect to Uvicorn using `curl --tlsv1.2 --tls-max 1.2`, probably Envoy ciphers are not compatible). Forcing Envoy to use 1.3 does the trick:

```yaml
common_tls_context:
  tls_params:
    tls_minimum_protocol_version: TLSv1_3
    tls_maximum_protocol_version: TLSv1_3
```

Hope this helps other people thay may come across this issue.

Thanks,

lorenzo
